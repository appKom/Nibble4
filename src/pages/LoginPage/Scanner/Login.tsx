import { handleRfid } from "api/authorization";
import React, {
  ChangeEvent,
  FC,
  FormEvent,
  useContext,
  useEffect,
  useRef,
  useState,
} from "react";
import styled from "styled-components";
import { Redirect } from "react-router-dom";
import { setUser } from "state/actions";
import { StateContext } from "state/state";
import { registerCardRoute } from "utility/routes";
import { modalTypes } from "types/modal";
import { setModalState } from "state/actions";

const Login: FC = () => {
  const inputRef = useRef<HTMLInputElement>(null);
  const { dispatch } = useContext(StateContext);
  const [registerCard, setRegisterCard] = useState(false);
  const [rfid, setRfid] = useState<string>("");

  const focusInput = () => {
    if (inputRef.current) inputRef.current.focus();
  };

  useEffect(() => {
    window.addEventListener("keydown", focusInput);
    return () => {
      window.removeEventListener("keydown", focusInput);
    };
  }, [inputRef]);

  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target;
    if (value.trim()) setRfid(value);
  };

  const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (!rfid.trim()) return;
    dispatch(setModalState(modalTypes.LOADING));
    const user = await handleRfid(rfid).then((user) => {
      dispatch(setModalState(modalTypes.DISABLED));
      return user;
    });
    if (!user) {
      dispatch(setModalState(modalTypes.FAILURE));
      return;
    }
    if (user.pk >= 0) dispatch(setUser(user));
    else setRegisterCard(true);
  };

  if (registerCard) return <Redirect to={registerCardRoute(rfid)} />;

  return (
    <form onSubmit={onSubmit}>
      <HiddenInput
        type="text"
        ref={inputRef}
        value={rfid}
        onChange={onChange}
      />
    </form>
  );
};

const HiddenInput = styled.input`
  display: block;
  transform: scale(0);
`;

export default Login;
