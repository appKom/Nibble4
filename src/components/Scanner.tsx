import React, { FC } from "react";
import styled from "styled-components";
import Login from "components/Login";

const Scanner: FC = () => (
  <Container>
    <h1>Logg inn</h1>
    <Login />
    <img src="images/rfid.png" alt="RFID icon" />
  </Container>
);

const Container = styled.div`
  margin: auto;
  h1 {
    text-align: center;
  }
  img {
    margin: auto;
    transform: scale(50%);
    max-width: 350px;
    display: flex;
  }
`;

export default Scanner;
