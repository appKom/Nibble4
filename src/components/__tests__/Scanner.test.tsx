import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Scanner from "../Scanner";

afterEach(cleanup);

it("renders", () => {
  const { asFragment } = render(<Scanner />);
  expect(asFragment()).toMatchSnapshot();
});
