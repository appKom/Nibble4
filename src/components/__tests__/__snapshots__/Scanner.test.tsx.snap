// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz dutjKU"
  >
    <h1>
      Logg inn
    </h1>
    <input
      class="sc-bdfBwQ bOBVXI"
      type="text"
      value=""
    />
    <img
      alt="RFID icon"
      src="images/rfid.png"
    />
  </div>
</DocumentFragment>
`;
